### ISSUE TITLE
Rewrite launchers to use a hybrid of scripting and C

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#11818]
As a system administrator, I'd like the ability to customize how Chapel launches programs by editing a script rather than C code in the runtime/ library and recompiling the launchers because it's more straightforward and approachable for my skillset and familiarity with Chapel.  As a Chapel user, I'd like there to be fewer bugs in the launchers because they're annoying to work around.  As a Chapel developer, I'd like an easier way to maintain common code across different launchers to keep them more maintainable and bug-free.

For some time, we've discussed a rewrite of the launchers in order to move the guts of them (particularly the construction and invocation of the `aprun`, `srun`, `gasnet_ibvrun`, etc. commands) into a script to make the launchers easier to edit, more like the scripts that they logically are, and potentially simpler to share code between.

The historical reason we haven't done this is that we need some compiler-generated C code to be executed at launch-time—most notably, the config parser so that (a) it will know about the user's config vars and consts and (b) it can catch any command-line errors prior to sitting in a queue for however long and eventually launching.  In addition, by having the launcher be compiler-generated, we can give it the executable name that the user expects and continue to shelter the beginning user from having to learn about the notion of running a program using a launcher script unrelated to their program name (e.g., `chplrun`) like the rest of HPC uses.

The proposed fix is to move as much of the launcher as possible / makes sense into scripts (Python or bash being the two most likely options) and having the C code invoke the scripts when it was time to launch them.

This epic proposes doing this conversion for the maintainability and customizability reasons given above.  The proposal would be to start with an easy case as a proof of concept, get it working as we want, and determine the interface(s) between C and the scripting language, and then propagate the same approach to other launchers as time permits (perhaps in order of importance / frequency of use / bugs found / frequency of customization by sites).

**Comment by ben-albrecht** [comment]
I would like to revisit some design aspects of this task soon, given it is a potential [GSoC 2019 project](https://chapel-lang.org/gsoc/ideas.html#launchers-rework).

The current leading proposal for this issue is to rewrite our launchers in Python. I think that would be *much* better than what we have today. However, I'd like to explore the option of rewriting out launchers in Chapel:

If our launcher was written in Chapel, we could produce a single binary, rather than 2 like we do today (`a.out` and `a.out_real`). The launching process could work by having the built program launch itself, where the login-node code path differs from the compute-node code path:

```bash
> export CHPL_LAUNCHER=slurm-srun
> chpl foo.chpl

# Only one binary is produced, rather than a separate launcher script
> ls
foo.chpl
foo

# foo was compiled with CHPL_LAUNCHER=slurm-srun, so
# it will launch itself with slurm by default
> ./foo -nl 4

# The Chapel program will create a subprocess to launch itself, where
# --launched=true skips launcher code:
#
#     srun -N 4 ./foo --launched=true

# User could also run on login node, the equivalent of running ./foo_real today
> ./foo -nl 1 --launched=true
```

I think this will be more intuitive for users than the current world of producing 2 separate binaries, or the python alternative where we produce a binary and python script.

In discussing this with @mppf, he mentioned that we could run into challenges in the future where we support cross-compiling on heterogeneous systems where say, a login node is x-86, and the compute nodes are arm. However, @mppf also mentioned that there were some possible solutions to this (I'll let him elaborate if he'd like to).

Additionally, this would allow arbitrary *chapel* code execution in a launcher, which could open the door for some interesting applications.

In writing the launchers in Chapel, we could expose the launcher modules to users, so that they could write complex job managers, where Chapel's parallelism features would be advantageous. @e-kayrakli brought this up in  https://github.com/chapel-lang/chapel/issues/10041:

> Everytime I need to do some experiment I end up writing set of python scripts that submits a ton of jobs, monitors them and collects the results (and plots them sometimes). Such scripts generally end up using python threads to do submission/monitoring/collection concurrently. Having such a module in Chapel (along with faster compilation times) would probably lead me to choose Chapel instead of Python.

Also in https://github.com/chapel-lang/chapel/issues/10041, @bradcray brought up some good criticisms of writing launchers in Chapel:

> (a) keeps something that seems "script-like" in a compiled code base, which adds overhead similar to the C code today

Agreed. The compile-times slowing down the edit-build-run cycle of modifying/adding launchers will be a downside of a Chapel-approach. I expect this won't be a big deal for Chapel users, who are used to Chapel compile-times, but it will be a pain point for sys-admins who have never used Chapel.

> (b) requires the sysadmins to use a language they're probably not familiar with;

This will definitely raise the barrier for modifying launchers, but I'm not sure how high this barrier will actually be. Python programmers tend to pick up Chapel quickly based on my anecdata, and the launcher programs tend to be fairly simple, limited to string and subprocess manipulation.

On the flip side, this may lower the barrier for Chapel users who need to modify/add launchers, since they are already writing Chapel in the first place and will not be required to know python.

> (c) doesn't make use of any Chapel-specific features (parallelism, locality, etc.).

I don't think this disqualifies Chapel for the job. Chapel is a general programming language after all.

----

I'd also like to point out a potential challenge with using python launchers - we are currently stuck continuing support for python 2.6, due to CLE 5 support. Therefore, our python launcher will have to be written to support python 2.6 and python 3+. Additionally, we will be limited to the python 2.6 standard library, since this won't be running in a virtual-env. Currently, the only user-facing python code we support like this is the `chplenv` module, which has been painful to maintain at times, e.g. we have had to back-port parts of python 3 standard library to python 2.6. These issues might be alleviated by dropping Chapel launcher support for CLE 5, e.g., or by generating virtualenv wrappers in addition to the python launcher code.

**Comment by bradcray** [comment]
I *really* don't like the idea of writing the launchers in Chapel rather than Python.  As Ben summarizes above, I view the goal of this issue as being to write a program that is scripting-like in nature in a true scripting language rather than a compiled one; and a familiar one so that sysadmins can customize it according to the idiosyncrasies of their system.

I'm also concerned about baking in the assumption that the same binary can be run on both the login and compute nodes without putting in the effort to make a multiple-ISA binary.  Though this is described as a potential concern for the future, I believe it's also a concern for some systems today. 

Moreover, it introduces coding complications like "if my module initialization code declares a huge distributed data structure, how do I prevent the launcher from initializing those modules and creating a data structure that it doesn't have the memory or locales to store?"  Yes, this can be worked around, but it feels like making a fairly simple piece of code ("please start running this program for me") far more complicated than it needs to be (to write, to support, to debug, to reason about).

While the notion of producing a single binary is attractive at some level, it's not the kind of thing I've heard many Chapel users complain about once they understand why two programs exist.  I also don't believe that the number of programs should go to three in this new launcher world, because I'd expect the Python script to be common across all compilations for a given launcher, and not created on a per-program basis (e.g., Chapel-generated binary invokes Python script to do the launching portion of the task).

In summary, writing the launchers in Python feels like adding needless work and complexity to a case that doesn't really require it (in my opinion) and where the hope was to go to something simpler, not more complex.

**Comment by bradcray** [comment]
[Note that it also seems like it would be very possible to have the compiler generate a single, multi-ISA binary without writing the launcher part of it in Chapel; it could just as easily be written in C and linked into the single-generated binary?  So if that's the real goal, I don't think that bringing "write the launcher in Chapel" into the mix is a necessary precondition.]

**Comment by gbtitus** [comment]
Even today, on a Cray XC with different login and compute node CPUs and a compilation that forced high levels of optimization, the resulting binary might not run on the login node.  You could get around this by only highly optimizing the target program part of the compile and doing a least-comm-denominator compilation on the code leading from the binary's start address through the launcher portions.  But that won't work for a hypothetical (and in the recent future, actual) system in which the login and compute nodes are totally different ISAs, for example x86_64 and ARM.

**Comment by mppf** [comment]
> If our launcher was written in Chapel, we could produce a single binary, rather than 2 like we do today (a.out and a.out_real)

> I'm also concerned about baking in the assumption that the same binary can be run on both the login and compute nodes without putting in the effort to make a multiple-ISA binary. 

I like the idea of 1 binary becoming a normal use case, but I don't think multiple-ISA binaries are really going anywhere on linux. See

https://en.wikipedia.org/wiki/Fat_binary#Linux

which says

>  In 2010, progress on FatELF had stopped, and developer Ryan Gordon declared FatELF to be dead

...



But if we just wanted to simplify the user experience, we might be able to achieve some similar goals by having a Chapel interpreter just-in-time compile programs on the machine they are running.

> [Note that it also seems like it would be very possible to have the compiler generate a single, multi-ISA binary without writing the launcher part of it in Chapel; it could just as easily be written in C and linked into the single-generated binary? So if that's the real goal, I don't think that bringing "write the launcher in Chapel" into the mix is a necessary precondition.]

> Even today, on a Cray XC with different login and compute node CPUs and a compilation that forced high levels of optimization, the resulting binary might not run on the login node.

I think it'd be nice if we could get into a place where only 1 binary is created in the event that the login node and the compute nodes share the same ISA. Then, that binary can have in the real `main` (before the Chapel user's main or the initialization code) the code to run the (Python) launcher if necessary, and that part of the program could be compiled unoptimized/unspecialized.

If the login node and compute nodes have different ISAs, I think we do need 2 binaries on Linux, at least based on the lack of support for multi-ISA binaries there.

> if my module initialization code declares a huge distributed data structure, how do I prevent the launcher from initializing those modules and creating a data structure that it doesn't have the memory or locales to store?

I think this is a good point and suggests that no matter what language the launcher is written in, we want it to run separately (i.e. even if it was written in Chapel, we wouldn't want it to run the launched program's module initialization functions; rather it would just run module initialization for the launcher itself).

> requires the sysadmins to use a language they're probably not familiar with;

I think this is a reasonable point and it begs the question - who is likely to need to modify the launchers? Chapel programmers? Chapel developers? Sysadmins?

I think if we expect it to be commonly sysadmins doing this, Python is a better choice.
